plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -356,6 +377,160 @@ private static ConnectorPageSource createOrcPageSource(
+
+        List<String> path = expectedDereferences.subList(readerDereferences.size(), expectedDereferences.size());
+        ImmutableList.Builder<Integer> dereferenceChain = ImmutableList.builder();
+        ColumnIdentity columnIdentity = readColumn.getColumnIdentity();
+        for (String dereference : path) {
+            List<ColumnIdentity> children = columnIdentity.getChildren();
+            for (int i = 0; i < children.size(); i++) {
+                if (children.get(i).getName().equals(dereference)) {
+                    dereferenceChain.add(i);
+                    columnIdentity = children.get(i);
+                    break;
+                }
+            }
+        }
+
+        return dereferenceChain.build();
+    }
+
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumnsForOrc(List<IcebergColumnHandle> columns)
+    {
+        // No projection is required if all columns are base columns
+        if (columns.stream().allMatch(IcebergColumnHandle::isBaseColumn)) {
+            return Optional.empty();
+        }
+
+        Map<IcebergColumnHandle, Integer> originalColumnIndexes = Streams.mapWithIndex(columns.stream(), Map::entry)
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> toIntExact(entry.getValue())));
+        // Order the columns in path depth order so existing projections for nested columns can be reused
+        List<IcebergColumnHandle> orderedColumns = originalColumnIndexes.keySet().stream()
+                .sorted(Comparator.comparingInt(col -> col.getPath().size()))
+                .collect(toImmutableList());
+
+        ImmutableList.Builder<ColumnHandle> projectedColumns = ImmutableList.builder();
+        Integer[] outputColumnMapping = new Integer[columns.size()];
+        Map<List<String>, Integer> mappedColumnIndices = new HashMap<>();
+        int projectedColumnIndex = 0;
+
+        for (IcebergColumnHandle column : orderedColumns) {
+            Optional<Integer> existingProjectedColumn = findExistingProjection(mappedColumnIndices, column);
+            int columnIndex = originalColumnIndexes.get(column);
+            if (existingProjectedColumn.isPresent()) {
+                outputColumnMapping[columnIndex] = existingProjectedColumn.get();
+                continue;
+            }
+
+            projectedColumns.add(column);
+            mappedColumnIndices.put(column.getFullPath(), projectedColumnIndex);
+            outputColumnMapping[columnIndex] = projectedColumnIndex;
+            projectedColumnIndex++;
+        }
+
+        return Optional.of(new ReaderColumns(projectedColumns.build(), ImmutableList.copyOf(outputColumnMapping)));
+    }
+
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumnsForParquet(List<IcebergColumnHandle> columns)
```

Comment:
What is the difference vs `projectColumnsForOrc`

Reply:
I meant to put this in the javadoc, but (IIUC) the Parquet reader expects columns to be referenced from the root of the file while Orc works fine referencing the nested columns directly. When I did Orc at first the ColumnHandle didn't need to keep track of the base column type, but for Parquet it does.